### ISSUE TITLE
Phi 3 Medium

### ISSUE THREAD

**Issue by unaveragetech** [unaveragetech/Gitbot#50]
What do you think of this new approach 

import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader
from transformers import AutoTokenizer, AutoModelForCausalLM
import os
import json
import numpy as np
from tqdm import tqdm
import pandas as pd
import glob
import pyarrow

# --- 1. Autoencoder Abstraction (Helper Class) ---
# Note: This requires the specific model code from 'thesephist/contra-bottleneck-t5-large-wikipedia'
# Ensure you have 'trust_remote_code=True' when loading if needed.

class BottleneckT5Autoencoder:
    def __init__(self, model_path: str, device='cpu'):
        self.device = device
        print(f"Using device: {self.device}")
        self.tokenizer = AutoTokenizer.from_pretrained(model_path, model_max_length=512)
        self.model = AutoModelForCausalLM.from_pretrained(model_path, trust_remote_code=True).to(self.device)
        self.model.eval()
        
    @torch.no_grad()
    def embed(self, text: str) -> torch.FloatTensor:
        """Encodes a single string into a latent embedding."""
        inputs = self.tokenizer(text, return_tensors='pt', truncation=True, max_length=512).to(self.device)
        decoder_input_ids = torch.tensor([[self.tokenizer.pad_token_id]], dtype=torch.long).to(self.device)
        outputs = self.model(
            input_ids=inputs['input_ids'],
            attention_mask=inputs['attention_mask'],
            decoder_input_ids=decoder_input_ids,
            encode_only=True,
        )
        return outputs

    @torch.no_grad()
    def embed_batch(self, texts: list[str]) -> torch.FloatTensor:
        """Encodes a batch of strings into latent embeddings."""
        inputs = self.tokenizer(
            texts, return_tensors='pt', padding=True, truncation=True, max_length=512
        ).to(self.device)
        decoder_start_token_id = self.model.config.decoder_start_token_id
        if decoder_start_token_id is None:
            decoder_start_token_id = self.tokenizer.pad_token_id
        decoder_input_ids = torch.full(
            (len(texts), 1), decoder_start_token_id, dtype=torch.long, device=self.device
        )
        outputs = self.model(
            **inputs, decoder_input_ids=decoder_input_ids, encode_only=True,
        )
        return outputs

    @torch.no_grad()
    def generate_from_latent(self, latent: torch.FloatTensor, max_length=512, temperature=0.4) -> str:
        """Decodes a latent embedding back into text."""
        if latent.dim() == 1:
            latent = latent.unsqueeze(0)
        latent = latent.to(self.device)
        output_sequences = self.model.generate(
            encoder_outputs=None,
            latent_vector=latent,
            max_length=max_length,
            do_sample=True,
            temperature=temperature,
            top_p=0.9,
            num_return_sequences=1,
            pad_token_id=self.tokenizer.eos_token_id
        )
        return self.tokenizer.decode(output_sequences, skip_special_tokens=True)

# --- Initialize the Autoencoder ---
if torch.cuda.is_available():
    device = 'cuda'
elif torch.backends.mps.is_available():
    device = 'mps'
else:
    device = 'cpu'
model_path = 'thesephist/contra-bottleneck-t5-large-wikipedia'
autoencoder = BottleneckT5Autoencoder(model_path=model_path, device=device)
print("Autoencoder loaded successfully.")

# --- 2. Data Preparation ---
# --- Combine Parquet files into a single JSONL (Example) ---
parquet_dir = "/path/to/your/parquet/files/"
output_jsonl_file = "/path/to/save/merged_output.jsonl"
parquet_files = glob.glob(os.path.join(parquet_dir, "train-*.parquet"))
print(f"Found Parquet files: {parquet_files}")
with open(output_jsonl_file, "a", encoding="utf-8") as outfile:
    for file_path in tqdm(parquet_files, desc="Processing Parquet files"):
        try:
            df = pd.read_parquet(file_path)
            print(f"Read {len(df)} rows from {os.path.basename(file_path)}")
            for _, row in tqdm(df.iterrows(), total=len(df), desc="Writing JSONL", leave=False):
                if 'instruction' in row and 'response' in row:
                    json_record = json.dumps({"instruction": row['instruction'], "response": row['response']})
                    outfile.write(json_record + '\n')
                else:
                    print(f"Skipping row due to missing columns: {row.to_dict()}")
        except Exception as e:
            print(f"Error processing {file_path}: {e}")
print(f"Finished merging Parquet files into {output_jsonl_file}")

# --- Generate Embeddings from JSONL ---
checkpoint_interval = 100_000
input_jsonl_file = "/path/to/save/merged_output.jsonl"
embeddings_file = "/path/to/save/embeddings.npy"
checkpoint_file = "/path/to/save/checkpoint.txt"
batch_size = 64

def process_batch(batch_texts_instructions, batch_texts_responses, autoencoder_model):
    if not batch_texts_instructions or not batch_texts_responses:
        return [], []
    try:
        instr_embeddings = autoencoder_model.embed_batch(batch_texts_instructions)
        resp_embeddings = autoencoder_model.embed_batch(batch_texts_responses)
        return instr_embeddings.cpu().numpy(), resp_embeddings.cpu().numpy()
    except Exception as e:
        print(f"Error processing batch: {e}")
        instr_np, resp_np = [], []
        for instr, resp in zip(batch_texts_instructions, batch_texts_responses):
            try:
                instr_emb = autoencoder_model.embed(instr).cpu().numpy()
                resp_emb = autoencoder_model.embed(resp).cpu().numpy()
                instr_np.append(instr_emb)
                resp_np.append(resp_emb)
            except Exception as item_e:
                print(f"Error processing item '{instr[:50]}...': {item_e}")
        return np.array(instr_np) if instr_np else [], np.array(resp_np) if resp_np else []

try:
    with open(input_jsonl_file, "r", encoding="utf-8") as f:
        total_lines = sum(1 for _ in f)
    print(f"Total lines in file: {total_lines}")
except FileNotFoundError:
    print(f"Error: Input JSONL file not found at {input_jsonl_file}")
    exit()
if os.path.exists(checkpoint_file):
    with open(checkpoint_file, "r") as f:
        last_processed_line = int(f.read().strip())
    print(f"Resuming from line: {last_processed_line + 1}")
else:
    last_processed_line = 0
    print("No checkpoint found. Starting from the beginning.")
if os.path.exists(embeddings_file) and last_processed_line > 0:
    try:
        existing_embeddings = np.load(embeddings_file)
        num_expected_embeddings = last_processed_line * 2
        if existing_embeddings.shape >= num_expected_embeddings:
            embeddings_list = list(existing_embeddings[:num_expected_embeddings])
            print(f"Loaded {len(embeddings_list)} embeddings from previous run (up to line {last_processed_line}).")
        else:
            print("Warning: Embedding file size doesn't match checkpoint. Starting embeddings list fresh.")
            embeddings_list = []
            last_processed_line = 0
    except Exception as e:
        print(f"Error loading existing embeddings: {e}. Starting fresh.")
        embeddings_list = []
        last_processed_line = 0
else:
    embeddings_list = []
if last_processed_line > 0:
    print("Warning: Checkpoint found but no embeddings file. Resetting checkpoint.")
    last_processed_line = 0
overall_line_count = last_processed_line
batch_instructions: list[str] = []
batch_responses: list[str] = []
try:
    with open(input_jsonl_file, "r", encoding="utf-8") as f:
        for _ in range(last_processed_line):
            next(f)
        pbar = tqdm(f, total=total_lines, initial=last_processed_line, unit="line", desc="Processing lines")
        for line in pbar:
            overall_line_count += 1
            try:
                obj = json.loads(line)
                if "instruction" in obj and "response" in obj:
                    batch_instructions.append(obj["instruction"])
                    batch_responses.append(obj["response"])
                else:
                    print(f"Skipping line {overall_line_count}: Missing 'instruction' or 'response'.")
                    continue
                if len(batch_instructions) == batch_size:
                    instr_np, resp_np = process_batch(batch_instructions, batch_responses, autoencoder)
                    for i in range(len(instr_np)):
                        embeddings_list.append(instr_np[i])
                        embeddings_list.append(resp_np[i])
                    batch_instructions = []
                    batch_responses = []
                if overall_line_count % checkpoint_interval == 0:
                    current_line_to_save = overall_line_count
                    with open(checkpoint_file, "w") as cf:
                        cf.write(str(current_line_to_save))
                    np.save(embeddings_file, np.array(embeddings_list, dtype=np.float32))
                    pbar.set_postfix_str(f"Checkpoint saved at line {current_line_to_save}")
            except json.JSONDecodeError:
                print(f"Skipping line {overall_line_count}: Invalid JSON.")
            except Exception as e:
                print(f"Error on line {overall_line_count}: {e}")
        if batch_instructions:
            instr_np, resp_np = process_batch(batch_instructions, batch_responses, autoencoder)
            for i in range(len(instr_np)):
                embeddings_list.append(instr_np[i])
                embeddings_list.append(resp_np[i])
        with open(checkpoint_file, "w") as cf:
            cf.write(str(overall_line_count))
        np.save(embeddings_file, np.array(embeddings_list, dtype=np.float32))
        print(f"Processing complete. Final line count: {overall_line_count}/{total_lines}. Embeddings saved.")
except FileNotFoundError:
    print(f"Error: Input JSONL file not found at {input_jsonl_file}")
except Exception as e:
    print(f"An unexpected error occurred: {e}")
    with open(checkpoint_file, "w") as cf:
        cf.write(str(overall_line_count - len(batch_instructions)))
    if embeddings_list:
        np.save(embeddings_file, np.array(embeddings_list, dtype=np.float32))
    print("Saved progress before exiting due to error.")

# --- 3. Latent Manipulator Model Definition ---
class LatentManipulator(nn.Module):
    """
    A Feed-Forward Network designed to manipulate latent embeddings.
    Takes a 1024-dim embedding and outputs a 1024-dim embedding.
    Uses intermediate layer outputs (choked) and concatenation for richness.
    """
    def __init__(self, dropout_rate=0.2):
        super(LatentManipulator, self).__init__()
        self.layer1 = nn.Sequential(nn.Linear(1024, 2048), nn.BatchNorm1d(2048), nn.LeakyReLU(0.01, inplace=True), nn.Dropout(dropout_rate))
        self.layer2 = nn.Sequential(nn.Linear(2048, 4096), nn.BatchNorm1d(4096), nn.LeakyReLU(0.01, inplace=True), nn.Dropout(dropout_rate))
        self.layer3 = nn.Sequential(nn.Linear(4096, 6144), nn.BatchNorm1d(6144), nn.LeakyReLU(0.01, inplace=True), nn.Dropout(dropout_rate))
        self.layer4 = nn.Sequential(nn.Linear(6144, 9216), nn.BatchNorm1d(9216), nn.LeakyReLU(0.01, inplace=True), nn.Dropout(dropout_rate))
        self.layer5 = nn.Sequential(nn.Linear(9216, 6144), nn.BatchNorm1d(6144), nn.LeakyReLU(0.01, inplace=True), nn.Dropout(dropout_rate))
        self.layer6 = nn.Sequential(nn.Linear(6144, 4096), nn.BatchNorm1d(4096), nn.LeakyReLU(0.01, inplace=True), nn.Dropout(dropout_rate))
        self.layer7 = nn.Sequential(nn.Linear(4096, 2048), nn.BatchNorm1d(2048), nn.LeakyReLU(0.01, inplace=True), nn.Dropout(dropout_rate))
        self.choke1 = nn.Sequential(nn.Linear(2048, 2048), nn.BatchNorm1d(2048), nn.LeakyReLU(0.01, inplace=True), nn.Dropout(dropout_rate))
        self.choke2 = nn.Sequential(nn.Linear(4096, 2048), nn.BatchNorm1d(2048), nn.LeakyReLU(0.01, inplace=True), nn.Dropout(dropout_rate))
        self.choke3 = nn.Sequential(nn.Linear(6144, 2048), nn.BatchNorm1d(2048), nn.LeakyReLU(0.01, inplace=True), nn.Dropout(dropout_rate))
        self.choke4 = nn.Sequential(nn.Linear(9216, 2048), nn.BatchNorm1d(2048), nn.LeakyReLU(0.01, inplace=True), nn.Dropout(dropout_rate))
        self.choke5 = nn.Sequential(nn.Linear(6144, 2048), nn.BatchNorm1d(2048), nn.LeakyReLU(0.01, inplace=True), nn.Dropout(dropout_rate))
        self.choke6 = nn.Sequential(nn.Linear(4096, 2048), nn.BatchNorm1d(2048), nn.LeakyReLU(0.01, inplace=True), nn.Dropout(dropout_rate))
        self.choke7 = nn.Sequential(nn.Linear(2048, 2048), nn.BatchNorm1d(2048), nn.LeakyReLU(0.01, inplace=True), nn.Dropout(dropout_rate))
        self.aLayer1 = nn.Sequential(nn.Linear(15360, 8192), nn.BatchNorm1d(8192), nn.LeakyReLU(0.01, inplace=True), nn.Dropout(dropout_rate))
        self.aLayer2 = nn.Sequential(nn.Linear(8192, 4096), nn.BatchNorm1d(4096), nn.LeakyReLU(0.01, inplace=True), nn.Dropout(dropout_rate))
        self.output_layer = nn.Linear(4096, 1024)

    def forward(self, x):
        x1 = self.layer1(x)
        x2 = self.layer2(x1)
        x3 = self.layer3(x2)
        x4 = self.layer4(x3)
        x5 = self.layer5(x4)
        x6 = self.layer6(x5)
        x7 = self.layer7(x6)
        c1 = self.choke1(x1)
        c2 = self.choke2(x2)
        c3 = self.choke3(x3)
        c4 = self.choke4(x4)
        c5 = self.choke5(x5)
        c6 = self.choke6(x6)
        c7 = self.choke7(x7)
        concat = torch.cat([x, c1, c2, c3, c4, c5, c6, c7], dim=1)
        out = self.aLayer1(concat)
        out = self.aLayer2(out)
        out = self.output_layer(out)
        return out

# --- Helper: Weight Initialization ---
def init_weights(m):
    """Applies Kaiming Normal initialization for LeakyReLU."""
    if isinstance(m, nn.Linear):
        nn.init.kaiming_normal_(m.weight, nonlinearity='leaky_relu')
        if m.bias is not None:
            nn.init.constant_(m.bias, 0)

# --- Helper: Count Parameters ---
def count_parameters(model):
    """Counts the number of trainable parameters in a model."""
    return sum(p.numel() for p in model.parameters() if p.requires_grad)
